core/xml.ts

```diff
@@ -19,42 +19,63 @@ import * as utilsXml from './utils/xml.js';
 import type {VariableModel} from './variable_model.js';
 import * as Variables from './variables.js';
 import type {Workspace} from './workspace.js';
-import {WorkspaceComment} from './workspace_comment.js';
-import {WorkspaceCommentSvg} from './workspace_comment_svg.js';
-import type {WorkspaceSvg} from './workspace_svg.js';
+import {WorkspaceSvg} from './workspace_svg.js';
 import * as renderManagement from './render_management.js';
+import {WorkspaceComment} from './comments/workspace_comment.js';
+import {RenderedWorkspaceComment} from './comments/rendered_workspace_comment.js';
+import {Coordinate} from './utils/coordinate.js';
 
 /**
  * Encode a block tree as XML.
  *
  * @param workspace The workspace containing blocks.
- * @param opt_noId True if the encoder should skip the block IDs.
+ * @param skipId True if the encoder should skip the block IDs. False by
+ *     default.
  * @returns XML DOM element.
  */
-export function workspaceToDom(
-  workspace: Workspace,
-  opt_noId?: boolean,
-): Element {
+export function workspaceToDom(workspace: Workspace, skipId = false): Element {
   const treeXml = utilsXml.createElement('xml');
   const variablesElement = variablesToDom(
     Variables.allUsedVarModels(workspace),
   );
   if (variablesElement.hasChildNodes()) {
     treeXml.appendChild(variablesElement);
   }
-  const comments = workspace.getTopComments(true);
-  for (let i = 0; i < comments.length; i++) {
-    const comment = comments[i];
-    treeXml.appendChild(comment.toXmlWithXY(opt_noId));
+  for (const comment of workspace.getTopComments()) {
+    treeXml.appendChild(
+      saveWorkspaceComment(comment as AnyDuringMigration, !skipId),
+    );
   }
   const blocks = workspace.getTopBlocks(true);
   for (let i = 0; i < blocks.length; i++) {
     const block = blocks[i];
-    treeXml.appendChild(blockToDomWithXY(block, opt_noId));
+    treeXml.appendChild(blockToDomWithXY(block, skipId));
   }
   return treeXml;
 }
 
+/** Serializes the given workspace comment to XML. */
+function saveWorkspaceComment(
+  comment: WorkspaceComment,
+  saveId = true,
+): Element {
+  const elem = utilsXml.createElement('comment');
+  if (saveId) elem.setAttribute('id', comment.id);
+
+  elem.setAttribute('x', `${comment.getRelativeToSurfaceXY().x}`);
```

Comment:
For `BlockSvg` this operation might be expensive because you have to look at a list of elements. Not sure how expensive it might be for ws comments in the current implementation, but might it be worth saving the value rather than calculate it twice in a row? If the values are cached in the class then it doesn't matter, but I haven't pulled down the branch to check

Reply:
Yeah this just returns the saved location, no expense.